Issues API: Unindexed field issue integration

coszio opened the pull request:
Needs: #3375. Tracked by: #3471 

Submits and solves the `UnindexedField` issue:
- submits when a search request takes longer than some threshold (currently 300ms)
- solves after a `CreateIndex` update is successful on the same collection and field name, or the collection is deleted

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

coszio committed:
submit an issue for unindexed field

coszio committed:
solve unindexed field issue when an index is created for the field

coszio committed:
also log a warning when issue is submitted

coszio committed:
fmt

coszio commented on lib/segment/src/index/struct_payload_index.rs:
```diff
@@ -309,7 +310,21 @@ impl StructPayloadIndex {
             Condition::Field(field_condition) => self
                 .estimate_field_condition(field_condition, nested_path)
-                .unwrap_or_else(|| CardinalityEstimation::unknown(self.available_point_count())),
+                .unwrap_or_else(|| {
+                    // Couldn't estimate cardinality, which means that there is no appropriate index
+                    // for the field, let's report an issue
+                    if let Ok(unindexed_field) = crate::problems::UnindexedField::try_from((
+                        field_condition.clone(),
+                        self.get_collection_name(),
+                    )) {
+                        let description = unindexed_field.description();
+                        if issues::submit(unindexed_field) {
+                            log::warn!("Performance issue: {}", description);
+                        }
+                    }
+
+                    CardinalityEstimation::unknown(self.available_point_count())
+                }),
```
For reviewers: Here is where this issue is triggered

coszio commented on src/common/points.rs:
```diff
@@ -547,14 +547,23 @@ pub async fn do_create_index_internal(
+        )
+        .await;
+
+    if res.is_ok() {
+        // We can deactivate issues related to this missing index
+        let code = segment::problems::UnindexedField::get_code(collection_name, &field_name);
+        issues::solve(code);
+    }
```
And here it is solved

coszio committed:
fix rebase problems

coszio committed:
get collection name without panicking

coszio committed:
better solving of issue

coszio committed:
remove outdated TODO

timvisee commented on lib/segment/src/problems/unindexed_field.rs:
```diff
@@ -0,0 +1,177 @@
+    ///
+    /// Will fail if the field condition cannot be used for inferring an appropriate schema.
+    /// For example, when there is no index that can be built to improve performance.
+    fn try_from(condition_collection_tuple: (FieldCondition, String)) -> Result<Self, Self::Error> {
```
I don't think I like this too much. I find converting `(FieldCondition, String)` into an `UnindexedField` type quite arbitrary.

To use it, you'd have to construct a tuple too.

What about just:

```rust
impl UnindexedField {
    fn new(condition: FieldCondition, collection: String) { ... }
}
```

timvisee commented on lib/storage/src/content_manager/toc/collection_meta_ops.rs:
```diff
@@ -58,7 +59,13 @@ impl TableOfContent {
+                let res = self.delete_collection(&operation.0).await;
+
+                // We can solve all issues related to this collection
+                if *res.as_ref().unwrap_or(&false) {
+                    issues::solve_by_filter(|code| code.starts_with(&operation.0));
+                }
```
Wouldn't it be better to just store the collection name in the issue, and match that directly.

I feel like matching part of a string here may be very error prone here.

For example, I could name my collection: `something.UNINDEXED_FIELD`

coszio replied:
I guess I didn't check our validation for collection names, so maybe we can use `/` as separator instead?

coszio committed:
add tests, fix GET /issues

coszio commented on lib/segment/src/index/struct_payload_index.rs:
```diff
@@ -309,7 +310,26 @@ impl StructPayloadIndex {
+                    {
+                        let description = unindexed_field.description();
+                        if issues::submit(unindexed_field) {
+                            log::warn!("Performance issue: {}", description);
```
it wasn't until recently that VSCode represented this formatting correctly, 'cause sometimes it wouldn't propagate squiggly lines there, so I got used to do it outside of the fmt string.

coszio committed:
review fixes

coszio committed:
use `/` instead of `.` as separator

coszio committed:
fmt

agourlay commented on lib/segment/src/index/struct_payload_index.rs:
```diff
@@ -309,7 +310,26 @@ impl StructPayloadIndex {
             Condition::Field(field_condition) => self
                 .estimate_field_condition(field_condition, nested_path)
-                .unwrap_or_else(|| CardinalityEstimation::unknown(self.available_point_count())),
+                .unwrap_or_else(|| {
```
TBH I am unsure if it is the best spot for generating the issues :thinking: 

The cardinality estimation is used to pick the driving filter from the query, that is the one with the highest selectivity.

I am worried this generates a lot of issue objects under high read load with lots of filters.
Most of those will be discarded because they exist already in the in-memory issues store. 

An alternative would be to place it in the function `condition_converter` to catch the actual field driving the query.

Can you maybe elaborate on why you picked this approach? :)

coszio replied:
While I agree that this might not be the best place for generating the issues. I would argue that it is more because it might have side effects for when the cardinality estimation is not directly used for searching or filtering, rather than if the issue is trying to be generated for every unindexed condition. 

My rationale is: if the search is slow already due to unindexed field, it won't affect much to try to do a read operation on the issues dashboard (when checking if the issue already exists).

After seeing if I can move this generation to the `condition_converter` function, I realized there is no easy way to include the collection name without drilling it down across many parent functions. So the question is: would it be worth it to increase the complexity for this?

coszio replied:
Another place we could add it is in:
https://github.com/qdrant/qdrant/blob/7ac7398bb129946c7c770f882272a0917d647237/lib/segment/src/index/struct_payload_index.rs#L456-L460

So that only the primary conditions trigger the issue

cc @generall

generall commented on lib/segment/src/index/struct_payload_index.rs:
```diff
@@ -331,6 +351,16 @@ impl StructPayloadIndex {
+        // Root dir -> storage -> collections -> ⭐️collection_name⭐️
+        self.path
+            .components()
+            .nth(3)
```
too ugly

coszio committed:
prepare extractor for time-based submission

coszio committed:
smol rename

coszio committed:
add check in search

coszio committed:
hook unindexed field creation in search and search_batch of rest

coszio committed:
update after rebase on `dev`

coszio commented:
Pending: do time-based issue generation only

coszio committed:
remove submit from `struct_payload_index`

coszio committed:
solve issues inside of handlers, not in match router

coszio committed:
post-process at Collection level

coszio committed:
don't clone filters

coszio committed:
remove from openapi spec

coszio committed:
- Add hidden env variable to adjust slow search threshold
- Remove Solution::None
- Fix UnindexedField issue extractor
- Add endpoint to openapi, but without response body spec
- Move integration test to consensus_tests to set lower threshold

coszio committed:
update num of apis

coszio commented:
Update: This issue is now triggered on core searches only, when the request takes more than 300ms (arbitrarily chosen) to fulfill. 

It scans the filter and checks against the current schema, so that, if the field is already indexed, it won't submit the issue.

The current `GET /issues` output for when there is a missing index on `sections` (collection name is `docs`), looks like this:
<img width="1186" alt="imagen" src="https://github.com/qdrant/qdrant/assets/62079184/dee2dbfe-3eb1-4976-a59e-41453b186350">

timvisee commented on lib/segment/src/problems/unindexed_field.rs:
```diff
@@ -0,0 +1,315 @@
+        })
+    }
+
+    pub fn get_code(collection_name: &str, field_name: &JsonPathV2) -> CodeType {
+        format!("/{collection_name}/UNINDEXED_FIELD/{field_name}")
+    }
```
Sorry, but I still don't like this string format because it  requires string based matching when automatically resolving issues. That's error prone.

Is there a good reason for not doing it in a strongly typed manner? Could that be a bad idea because various issues might be on wildly different levels, where a simple string could easily represent them all?

How about a flat structure where all issues are on the same level, having some constant type such as `UNINDEXED_FIELD` where each type can have their own respective metadata - a collection and field name in this case?

coszio replied:
I kinda share your thoughts on this... 

How about this: If we adhere to a REST-like representation, we can avoid headaches. In this case, a better code for this issue could be `/collections/{collection_name}/points/index/{field_name}/UNINDEXED_FIELD`

What makes me reluctant to have a flat structure is that it will want to have a different implementation of the `issues::Dashboard`, in which it becomes event-based, so that we don't pollute the codebase too much by separating the event from the handling. Maybe it is not a bad thing either, but it will require at least another PR.

coszio commented:
Closing in favor of #4139
